ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Util.hs

```diff
@@ -74,20 +75,32 @@ wrapFsError :: Monad m
 wrapFsError fsErr volDBErr action =
     tryVolDB fsErr volDBErr action >>= either (throwError volDBErr) return
 
--- Throws an error if one of the given file names does not parse.
+-- | Parses the 'FsPath' of each filename and zips them together.
+-- If any 'FsPath' does not parse, we return an error.
+parseAllFds :: forall blockId.
+               [FsPath]
+            -> Either (VolatileDBError blockId) [(FileId, FsPath)]
+parseAllFds = mapM go
+  where
+    go :: FsPath -> Either (VolatileDBError blockId) (FileId, FsPath)
+    go file = case parseFd file of
+      Nothing -> Left $ UnexpectedError $ ParserError $ InvalidFilename file
+      Just fd -> Right (fd, file)
+
+-- | Throws an error if one of the given file names does not parse.
 findLastFd :: forall blockId.
               Set FsPath
            -> Either (VolatileDBError blockId) (Maybe FileId)
-findLastFd files = foldM go Nothing files
-    where
-        maxMaybe :: Ord a => Maybe a -> a -> a
-        maxMaybe ma a = case ma of
-            Nothing -> a
-            Just a' -> max a' a
-        go :: Maybe FileId -> FsPath -> Either (VolatileDBError blockId) (Maybe FileId)
-        go fd file = case parseFd file of
-            Nothing  -> Left $ UnexpectedError $ ParserError $ InvalidFilename file
-            Just fd' -> Right $ Just $ maxMaybe fd fd'
+findLastFd = foldM go Nothing
+  where
+    maxMaybe :: Ord a => Maybe a -> a -> a
+    maxMaybe ma a = case ma of
+      Nothing -> a
+      Just a' -> max a' a
+    go :: Maybe FileId -> FsPath -> Either (VolatileDBError blockId) (Maybe FileId)
```

Comment:
Hmm, this is basically the same `go` as the one in `parseAllFds`. Rename it to `parseFd`, pull out the `maxMaybe`, and promote it to the top-level so it can be shared between `parseAllFds` and `findLastFd`.

Ah, I see `parseFd` already exists. I would unify the two:
```haskell
parseFd :: FsPath -> Either (VolatileDBError blockId) FileId
parseFd file = maybe err Right $
    parseFilename <=< lastMaybe $ fsPathToList file
  where
    err = Left $ UnexpectedError $ ParserError $ InvalidFilename file

    parseFilename :: Text -> Maybe FileId
    parseFilename = readMaybe
                  . T.unpack
                  . snd
                  . T.breakOnEnd "-"
                  . fst
                  . T.breakOn "."
```

Update `unsafeParseFd` by using `either .. id ..`, 
`parseAllFds = mapM $ \f -> (,f) <$> parseFd f`, and
`findLastFd = fmap safeMaximum . mapM parseFd . Set.toList`.